### ISSUE TITLE
provide Orientation value for RAW images

### ISSUE THREAD

**Issue by kfjahnke** [AcademySoftwareFoundation/OpenImageIO#4152]
As requested, this is a split-off from #4123.
As posted there, I think that OIIO should handle image orientation like this:

- if autorotation is available, use it per default and yield 'Orientation' = 1, but
- allow the user to switch autorotation off (no effect if it can't be had in the first place)
- if no autorotation was performed, provide an EXIF-compatible 'Orientation' value

This should apply to RAW images as well. So for raw images, the default could remain as it is (passing raw:user_flip=-1 to the plugin). If raw::user_flip is passed zero (no autorotation), the data should come in as recorded by the sensor, and the OIIO 'Orientation' property should be set to the corresponding EXIF value - I think it would be wise to adhere to this as the uniform standard, rather than using a second set of numerical values just for RAW files.
If the user passes any other value for raw:user_flip, the data would be pre-rotated by the plugin, and the OIIO 'Orientation' property should be set to 1, to reflect that the data - as imported - have been rotated under the user's command - if the user passes in an inappropriate raw:user_flip that's their mistake (or may be deliberate).
If finding out the camera orientation is relevant, users can pass raw:user_flip=0 and glean the orientation from the 'Orientation' property, which in turn should be available via the raw:flip value with subsequent translation from libraw numerical codes to EXIF standard. Then they can decide how they want to have the file read.
It might be a good idea to add a global OIIO flag to deliberately switch autorotation off for all plugins which support it. This would save users the hassle of coding per-format parameterization if they need to handle orientation themselves.

**Comment by lgritz** [comment]
Agree on all your points. I'm tagging this as a "help wanted" issue and for anybody reading who knows our secret codeword:  🎣 Mackerel!

**Comment by kfjahnke** [comment]
> for anybody reading who knows our secret codeword: 🎣 Mackerel!

Would you like to share what your secret codeword means?

>  assuming that oiio: arguments should override raw: arguments - should they?

I suppose this is a policy issue. What's the policy if there is a global oiio: argument and a plugin argument which affect the same feature? Should the plugin still be allowed to accept it's own brand of argument or should it give up on it and react to the oiio: argument instead? The point in case here being oiio:reorient and raw:user_flip: the latter would seem largely redundant if the plugin were to react to oiio:reorient, but beyond not touching or autorotating the image it has the extra feature to orient to a given orientation value whereas oiio:reorient is a boolean.

**Comment by lgritz** [comment]
> > for anybody reading who knows our secret codeword: 🎣 Mackerel!
> 
> Would you like to share what your secret codeword means?

Yes! It means: "here's a task that's would be ideal for somebody who has been looking for a way to get more involved with the project."  It also means "this is something that somebody other than Larry can do, so it's really best for the project if somebody other than Larry does it."

We wanted a catchy, quick way to say that, meant to invoke the old saying, "give a person a fish and you feed them for a day, but teach them to fish and you feed them for a lifetime."

**Comment by lgritz** [comment]
> I suppose this is a policy issue. What's the policy if there is a global oiio: argument and a plugin argument which affect the same feature? Should the plugin still be allowed to accept it's own brand of argument or should it give up on it and react to the oiio: argument instead? The point in case here being oiio:reorient and raw:user_flip: the latter would seem largely redundant if the plugin were to react to oiio:reorient, but beyond not touching or autorotating the image it has the extra feature to orient to a given orientation value whereas oiio:reorient is a boolean.

I think raw:reorient made sense if it was just meant for raw to control things in a very specific way for that reader. It should be changed in documentation to oiio:reorient if we expect a command that works the same way on many/all file types. We can still quietly accept raw:reorient as a synonym for compatibility.

**Comment by ssh4net** [comment]
> This should apply to RAW images as well. So for raw images, the default could remain as it is (passing raw:user_flip=-1 to the plugin). If raw::user_flip is passed zero (no autorotation), the data should come in as recorded by the sensor, and the OIIO 'Orientation' property should be set to the corresponding EXIF value - I think it would be wise to adhere to this as the uniform standard, rather than using a second set of numerical values just for RAW files.

in Libraw, orientation works this way:
-1 - rotation took from EXIF orientation tag.
0-8 - EXIF orientation tag replaced with user defined value.

0 == 0-rotation (usually horizontal orientation for a camera with a viewfinder on top).

If I remember right, OIIO in the libraw plugin inherits this behavior. But forcing Libraw plugin to use a user orientation required preload raw file header.

**Comment by lgritz** [comment]
I suspect this is long since fixed, so closing this stale ticket. Re-open or re-file if there is still a deficiency in orientation for raw.
